File: Python/Graph/bfs.py
from collections import deque


class Graph:
    def __init__(self, edges):
        self.adjList = {}

        for u, v in edges:
            if u in self.adjList:
                self.adjList[u].append(v)
            else:
                self.adjList[u] = [v]

    def bfs(self, source):
        q = deque()
        q.append(source)
        visited = [source]

        while q:
            current = q.popleft()
            print(current, end=' ')

            for node in self.adjList[current]:
                if node not in visited:
                    visited.append(node)
                    q.append(node)

    def dfs(self, source, visited=None):
        if visited is None:
            visited = []
        visited.append(source)
        print(source, end=' ')

        for node in self.adjList[source]:
            if node not in visited:
                self.dfs(node, visited)


if __name__ == '__main__':
    edges = [
        (1, 2),
        (1, 3),
        (2, 1),
        (2, 4),
        (2, 5),
        (3, 1),
        (3, 4),
        (3, 6),
        (4, 2),
        (4, 3),
        (4, 5),
        (4, 6),
        (5, 2),
        (5, 4),
        (5, 6),
        (6, 3),
        (6, 4),
        (6, 5),
    ]

    graph = Graph(edges)
    print(graph.adjList)

    graph.bfs(1)
    print()
    graph.dfs(1)

File: C++/Graph/bfs.cpp
#include <bits/stdc++.h>

using namespace std;

class Graph
{
private:
    int numberOfVertex;
    vector<vector<int>> adjList;

public:
    Graph(int numberOfVertex)
    {
        this->numberOfVertex = numberOfVertex;
        adjList.resize(numberOfVertex);
    }

    void addEdge(int source, int destination)
    {
        adjList[source].push_back(destination);
        adjList[destination].push_back(source);
    }

    void print()
    {
        for (int i = 0; i < adjList.size(); i++)
        {
            cout << i << " -> ";

            for (int j = 0; j < adjList[i].size(); j++)
            {
                cout << adjList[i][j] << " ";
            }

            cout << endl;
        }
    }

    void bfs(int source)
    {
        vector<bool> visited(numberOfVertex, false);

        visited[source] = true;
        queue<int> q;
        q.push(source);

        while (!q.empty())
        {
            int data = q.front();
            q.pop();

            cout << data << " -> ";

            for (auto neighbor : adjList[data])
            {
                if (!visited[neighbor])
                {
                    visited[neighbor] = true;
                    q.push(neighbor);
                }
            }
        }
    }
};

int main()
{
    int n, e;

    cin >> n >> e;

    Graph graph = Graph(n);

    for (int i = 0; i < e; i++)
    {
        int u, v;
        cin >> u >> v;

        graph.addEdge(u, v);
    }

    graph.print();

    graph.bfs(5);

    return 0;
}

/*
v e
6 8

0 1
0 2
0 4
1 4
2 3
3 4
3 5
4 5
*/